### PR TITLE
math: trig: Improve trig function execution speed

### PR DESCRIPTION
**Pull request by ShriramShastry**
if conditions in loops slow down the execution speed.So
 to optimize the loop containing if conditions where branch
 is separated by sign only. A new arithmetic math added .

 Added new defines to replace Q_CONVERT_FLOAT constants.

Signed-off-by: ShriramShastry <malladi.sastry@intel.com>

### CODE REVIEW

**Comment by lgirdwood** [comment]
@ShriramShastry how do you know this is faster ? Do you have FW logs to show the difference ?

File: src/math/trig.c [hunk 1]
```diff
@@ -58,16 +53,16 @@ void cordic_approx(int32_t th_rad_fxp, int32_t a_idx, int32_t *sign, int32_t *b_
 	 * without overflow.Increase of fractionLength makes the addition or
 	 * subtraction of a multiple of pi/2 more precise
 	 */
-	if (th_rad_fxp > cord_sincos_piovertwo_q28fl) {
-		if ((th_rad_fxp - cord_sincos_piovertwo_q29fl) <= cord_sincos_piovertwo_q28fl) {
-			th_rad_fxp -= cord_sincos_piovertwo_q29fl;
+	if (th_rad_fxp > PIOVERTWO_Q4_28) {
+		if ((th_rad_fxp - PIOVERTWO_Q3_29) <= PIOVERTWO_Q4_28) {
```
Comment (johnylin76) [inline src/math/trig.c:57]: I have an idea which might reduce some calculations: replace the statement
        (th_rad_fxp - PIOVERTWO_Q3_29) <= PIOVERTWO_Q4_28
by
        th_rad_fxp <= THREEQUARTERSOFPI_Q3_29
where THREEQUARTERSOFPI_Q3_29 = (PIOVERTWO_Q3_29 + PIOVERTWO_Q4_28) and also a constant

File: src/math/trig.c [hunk 2]
```diff
@@ -58,16 +53,16 @@ void cordic_approx(int32_t th_rad_fxp, int32_t a_idx, int32_t *sign, int32_t *b_
...
+			th_rad_fxp -= PIOVERTWO_Q3_29;
 			*sign  = -1;
 		} else {
 			th_rad_fxp -= cordic_sine_cos_piovertwo_q30fl;
 		}
-	} else if (th_rad_fxp < -cord_sincos_piovertwo_q28fl) {
-		if ((th_rad_fxp + cord_sincos_piovertwo_q29fl) >= -cord_sincos_piovertwo_q28fl) {
-			th_rad_fxp += cord_sincos_piovertwo_q29fl;
+	} else if (th_rad_fxp < -PIOVERTWO_Q4_28) {
+		if ((th_rad_fxp + PIOVERTWO_Q3_29) >= -PIOVERTWO_Q4_28) {
```
Comment (johnylin76) [inline src/math/trig.c:64]: ditto

**Comment by ShriramShastry** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@ShriramShastry is this still draft and pls remember to add your perfroamcne data to the commit message. The Zephyr build failure is unrelated to your PR.

**Comment by ShriramShastry** [comment]
@lgirdwood , I will soon move it to review state. I am using Xtensa Xplorer for preparation work. In the moment, I am capturing the CPU cycles with and without optimization . Sure, I will add performance data to commit message.

**Review by singalsu — APPROVED** [review]

File: src/math/trig.c [hunk 3]
```diff
@@ -168,7 +158,7 @@ int32_t is_scalar_cordic_acos(int32_t cosvalue, int16_t numiters)
 			y += ydshift;
 		} else {
 			sign = (((x > cosvalue) && (y >= 0)) ||
-				((x < cosvalue) && (y < 0))) ? 1 : -1;
+				((x < cosvalue) && (y < 0))) * 2 - 1;
```
Comment (cujomalainey) [inline src/math/trig.c:161]: multiplication is faster than a ternary?

Comment (ShriramShastry) [reply src/math/trig.c:161]: Yes, a bit faster !!

**Comment by johnylin76** [comment]
Same as [PR#4502](https://github.com/thesofproject/sof/pull/4502#issuecomment-914166730), I have run the measurement of CPU ticks with and without this PR commit on TGL by perf_cnt tool. Here are the results:

CPU ticks per sin() Implementation
335.78 Cordic 16bit generic
448.36 Cordic 16bit generic + this PR commit

From this perspective the speed is not improved but getting even worse.

**Comment by ShriramShastry** [comment]
> Same as [PR#4502](https://github.com/thesofproject/sof/pull/4502#issuecomment-914166730), I have run the measurement of CPU ticks with and without this PR commit on TGL by perf_cnt tool. Here are the results:
> 
> CPU ticks per sin() Implementation
> 335.78 Cordic 16bit generic
> 448.36 Cordic 16bit generic + this PR commit
> 
> From this perspective the speed is not improved but getting even worse.

Thanks for the measurement !!Can you please share GitHub legacy `LUT solution sin_fixed()` CPU ticks ?

**Comment by johnylin76** [comment]
It

> > Same as [PR#4502](https://github.com/thesofproject/sof/pull/4502#issuecomment-914166730), I have run the measurement of CPU ticks with and without this PR commit on TGL by perf_cnt tool. Here are the results:
> > CPU ticks per sin() Implementation
> > 335.78 Cordic 16bit generic
> > 448.36 Cordic 16bit generic + this PR commit
> > From this perspective the speed is not improved but getting even worse.
> 
> Thanks for the measurement !!Can you please share GitHub legacy `LUT solution sin_fixed()` CPU ticks ?

I shared in [PR#4502](https://github.com/thesofproject/sof/pull/4502#issuecomment-914166730). Paste here as well.

CPU ticks per sin() Implementation
69.92 Legacy LUT
613.66 Cordic 32bit generic
335.78 Cordic 16bit generic
448.36 Cordic 32bit HiFi3
217.96 Cordic 16bit HiFi3

**Comment by ShriramShastry** [comment]
> 69.92 Legacy LUT

Thanks !! 69.92 Legacy LUT is not inline with my observation. We will align once you are back from your vacation.

**Comment by ShriramShastry** [comment]
> It
> 
> > > Same as [PR#4502](https://github.com/thesofproject/sof/pull/4502#issuecomment-914166730), I have run the measurement of CPU ticks with and without this PR commit on TGL by perf_cnt tool. Here are the results:
> > > CPU ticks per sin() Implementation
> > > 335.78 Cordic 16bit generic
> > > 448.36 Cordic 16bit generic + this PR commit
> > > From this perspective the speed is not improved but getting even worse.
> > 
> > 
> > Thanks for the measurement !!Can you please share GitHub legacy `LUT solution sin_fixed()` CPU ticks ?
> 
> I shared in [PR#4502](https://github.com/thesofproject/sof/pull/4502#issuecomment-914166730). Paste here as well.
> 
> CPU ticks per sin() Implementation
> 69.92 Legacy LUT
> 613.66 Cordic 32bit generic
> 335.78 Cordic 16bit generic
> 448.36 Cordic 32bit HiFi3
> 217.96 Cordic 16bit HiFi3

The differences in CPU cycles is due to core configuration. With `TGL Core` `Legacy LUT` is better while `Cordic 16bit` is good  with `generic core`

**Comment by lgirdwood** [comment]
> > CPU ticks per sin() Implementation
> > 69.92 Legacy LUT
> > 613.66 Cordic 32bit generic
> > 335.78 Cordic 16bit generic
> > 448.36 Cordic 32bit HiFi3
> > 217.96 Cordic 16bit HiFi3
> 
> Thanks for the alignment ,To summaries the observations,
> The differences in CPU cycles is due to core configuration. With `TGL Core` `Legacy LUT` is better while `Cordic 16bit` is good with `generic core`

So whats the next steps here ? I assume legacy does not provide the same audio quality vs cordic sine ?

1) Do we need to further optimise.
2) Do we need a kconfig to select cordic or legacy ?

@singalsu what's your take ?

**Comment by singalsu** [comment]
@lgirdwood Legacy is more accurate than 16 bit cordic but not quite as accurate as 32 bit.  I set the LUT size & interpolation there to match about 20 bits audio. I think we need to bring back the legacy as kconfig option. With intrinsics it would be even faster than previous pure C for the interpolation part. I can make a PR next week.

Also I plan to try another shot with intrinsics for cordic. It could be the default at least for small memory limited platforms. But LUT sine will remain fastest.

**Comment by lgirdwood** [comment]
> @lgirdwood Legacy is more accurate than 16 bit cordic but not quite as accurate as 32 bit. I set the LUT size & interpolation there to match about 20 bits audio. I think we need to bring back the legacy as kconfig option. With intrinsics it would be even faster than previous pure C for the interpolation part. I can make a PR next week.
> 
> Also I plan to try another shot with intrinsics for cordic. It could be the default at least for small memory limited platforms. But LUT sine will remain fastest.

Ok, this make sense if it's all under Kconfig. Users can choose "slower but smaller" or "larger but faster" in the Kconfig. This gives some flexibility with memory and MIPS.

**Comment by ShriramShastry** [comment]
Replaced with new PR https://github.com/thesofproject/sof/pull/4942

**Pull request closed without merging**